plugin/src/main/scala/migrate/ScalacOptionsMigration.scala

```diff
@@ -1,123 +1,94 @@
-    val sv            = scalaVersion.value
-
-    if (!sv.startsWith("2.13."))
+
+  lazy val internalImpl: Def.Initialize[Task[Unit]] = Def.task {
+    val logger           = Keys.streams.value.log
+    val projectId        = Keys.thisProject.value.id
+    val allScalacOptions = allScalacOptionsTask.value
+    val sv               = Keys.scalaVersion.value
+
+    if (!sv.startsWith("2.13.") && !sv.startsWith("3."))
       throw new MessageOnlyException(Messages.notScala213(sv, projectId))
 
-    logger.info(starting(projectId, configs.map(_.id)))
-    logger.info(warning)
-    logger.info(help)
+    logger.info(startingMessage(projectId))
+
+    val migrated = migrateAPI.migrateScalacOption(allScalacOptions.asJava)
 
-    if (commonOptions.nonEmpty) {
-      val migrationStatus = migrateAPI.migrateScalacOption(commonOptions.asJava)
-      reportStatus(logger, migrationStatus)
+    if (migrated.getValid.nonEmpty) {
+      logger.info(validMessage(migrated.getValid))
     }
 
-    Def.task {
-      val logger        = streams.value.log
-      val configOptions = configs.map(_ / scalacOptions).join.value
-      for {
-        (options, config) <- configOptions.zip(configs)
-        filteredOptions    = options.filterNot(opt => commonOptions.contains(opt))
-        if filteredOptions.nonEmpty
-      } {
-        logger.info(s"${BOLD}In configuration ${config.id}:${RESET}")
-        val migrationStatus = migrateAPI.migrateScalacOption(filteredOptions.asJava)
-        reportStatus(logger, migrationStatus)
-      }
+    val renamed = migrated.getRenamed.asScala.toMap
+    if (renamed.nonEmpty) {
+      logger.warn(renamedMessage(renamed))
     }
-  }
 
-  private def reportStatus(logger: ManagedLogger, status: ScalacOptions): Unit = {
-    val notParsed = status.getNotParsed.toSeq
-    if (notParsed.nonEmpty) {
-      logger.warn(s"""|
-                      |We were not able to parse the following ScalacOptions:
-                      |${formatScalacOptions(notParsed)}
-                      |
-                      |""".stripMargin)
+    if (migrated.getRemoved.nonEmpty) {
+      logger.warn(removedMessage(migrated.getRemoved))
     }
 
-    val specific2 = status.getSpecificScala2.toSeq
-    val scala3    = status.getScala3cOptions.toSeq
-    val renamed   = status.getRenamed.asScala.toMap
-    val plugins   = status.getPluginsOptions.toSeq
-    logger.info(message(specific2, renamed, scala3, plugins))
+    if (migrated.getUnknown.nonEmpty) {
+      logger.warn(unknownMessage(migrated.getUnknown))
+    }
   }
 
-  private def starting(projectId: String, configs: Seq[String]): String =
+  private val allScalacOptionsTask: Def.Initialize[Task[Seq[String]]] = Def.taskDyn {
+    val configs = migrationConfigs.value
+    Def.task {
+      configs
+        .map(c => (c / Keys.scalacOptions).result)
+        .join(_.join)
+        .value
+        .collect { case Value(scalacOptions) => scalacOptions }
+        .flatten
+        .distinct
```

Comment:
The `.distinct` seems dangerous at this level. Since some options are spread over two elements of the `scalacOptions`, like `"-encoding", "utf-8"`, doing a `distinct` at the level of each element is risky. What if there is another option that also takes `"utf-8"` as "parameter"?

IMO it would be safer to do the `.distinct` later, having having `sanitize`d the options.